plugins/autosuspend.koplugin/main.lua

```diff
@@ -43,90 +48,90 @@ function AutoSuspend:_enabled()
     return self.auto_suspend_sec > 0
 end
 
-function AutoSuspend:_schedule(settings_id)
+function AutoSuspend:_enabledShutdown()
+    return Device:canPowerOff() and self.autoshutdown_sec > 0
+end
+
+function AutoSuspend:_schedule()
     if not self:_enabled() then
         logger.dbg("AutoSuspend:_schedule is disabled")
         return
     end
-    if self.settings_id ~= settings_id then
-        logger.dbg("AutoSuspend:_schedule registered settings_id ",
-                   settings_id,
-                   " does not equal to current one ",
-                   self.settings_id)
-        return
-    end
 
-    local delay
+    local delay_suspend, delay_shutdown
 
     if PluginShare.pause_auto_suspend then
-        delay = self.auto_suspend_sec
+        delay_suspend = self.auto_suspend_sec
+        delay_shutdown = self.autoshutdown_sec
     else
-        delay = self.last_action_sec + self.auto_suspend_sec - os.time()
+        delay_suspend = self.last_action_sec + self.auto_suspend_sec - os.time()
+        delay_shutdown = self.last_action_sec + self.autoshutdown_sec - os.time()
     end
 
-    if delay <= 0 then
+    if delay_suspend <= 0 then
         logger.dbg("AutoSuspend: will suspend the device")
         UIManager:suspend()
+    elseif delay_shutdown <= 0 then
+        logger.dbg("AutoSuspend: initiating shutdown")
+        UIManager:poweroff_action()
     else
-        logger.dbg("AutoSuspend: schedule at ", os.time() + delay)
-        UIManager:scheduleIn(delay, function() self:_schedule(settings_id) end)
+        if self:_enabled() then
+            logger.dbg("AutoSuspend: schedule suspend at ", os.time() + delay_suspend)
+            UIManager:scheduleIn(delay_suspend, self._schedule, self)
+        end
+        if self:_enabledShutdown() then
+            logger.dbg("AutoSuspend: schedule shutdown at ", os.time() + delay_shutdown)
+            UIManager:scheduleIn(delay_shutdown, self._schedule, self)
+        end
     end
 end
 
-function AutoSuspend:_deprecateLastTask()
```

Comment:
good riddance:)

Reply:
This was convoluted and hard to follow. The changed method Just Works™ (thanks to the scheduleIn/unschedule methods I wrote in 2014 I might add with some pride).